src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs

```diff
@@ -94,5 +97,152 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
                 return null;
             }
         }
+
+        /// <summary>
+        /// Visits the binary operator tree of interpolated string additions in a depth-first in-order visit.
+        /// <paramref name="stringCallback"/> controls whether to continue the visit by returning true or false:
+        /// if true, the visit will continue. If false, the walk will be cut off.
+        /// </summary>
+        public static bool VisitBinaryOperatorInterpolatedString<TInterpolatedStringType, TArg>(
+            this BoundBinaryOperator binary,
+            TArg arg,
+            Func<TInterpolatedStringType, TArg, bool> stringCallback,
+            Action<BoundBinaryOperator, TArg>? binaryOperatorCallback = null)
+            where TInterpolatedStringType : BoundInterpolatedStringBase
+        {
+            var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
+
+            pushLeftNodes(binary, stack);
+
+            while (stack.TryPop(out BoundBinaryOperator? current))
+            {
+                switch (current.Left)
+                {
+                    case BoundBinaryOperator:
+                        break;
+                    case TInterpolatedStringType interpolatedString:
+                        if (!stringCallback(interpolatedString, arg))
+                        {
+                            return false;
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(current.Left.Kind);
+                }
+
+                binaryOperatorCallback?.Invoke(current, arg);
```

Comment:
>binaryOperatorCallback?.Invoke(current, arg); [](http://example.com/codeflow?start=16&length=45)

Invoking a call back for the parent before visting right child doesn't quite match what I expect from a depth-first visit.

Reply:
Left, middle, right is what defines a depth-first in-order visit, which is what is documented for this function to perform.

Reply:
Is this order chosen just to avoid a recursion on the right? In that case, I would not optimize for that scenario and stick to the order (left, right, middle) as, I believe,  we usually do. We usually do not visit parent node in between visting different children.

Reply:
This order is chosen to ensure that speculation in the nullable walker will follow a lexical ordering. IE, if you put your position in the middle of a `+`, the speculative model will give you the correct result of the left having been evaluated, but not the right.

Reply:
> if you put your position in the middle of a +, the speculative model will give you the correct result of the left having been evaluated, but not the right.

Could you please elaborate how exactly is this achieved?

Reply:
I went to explain how it would be achieved, then realized that isn't how it would work, because I was misremembering where the `+` syntax start is. I'll switch to a pre-order visit.